src/DefinitionResolver.php

```diff
@@ -479,8 +483,10 @@ public function resolveExpressionNodeToType(Node\Expr $expr): Type
                 if ($expr instanceof Node\Expr\MethodCall) {
                     $fqn .= '()';
                 }
-                $def = $this->index->getDefinition($fqn);
-                if ($def !== null) {
+                if ($def = $this->index->getDefinition($fqn)) {
+                    if ($def->type instanceof Types\This || $def->type instanceof Types\Self_) {
+                        return $this->resolveExpressionNodeToType($expr->var);
+                    }
```

Comment:
It is documented here: https://www.phpdoc.org/docs/latest/guides/types.html#keywords

Yes, it is needed for fluent interfaces, I just thought you would like to merge in a solution that fixes $this autocompletion without support for fluent interfaces first and then do a follow-up